modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -145,13 +145,282 @@ class SoftMaxSubgraph : public Subgraph
+                scaleW = (int)nodes_attributes[3].at<float>(0);
+            }
+            else
+            {
+                scaleH = (int)nodes_attributes[3].at<float>(0);
+                scaleW = (int)nodes_attributes[1].at<float>(0);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    virtual void finalize(const Ptr<ImportGraphWrapper>&,
+                          const Ptr<ImportNodeWrapper>& fusedNode,
+                          std::vector<Ptr<ImportNodeWrapper> >& inputs) CV_OVERRIDE
+    {
+        opencv_onnx::NodeProto* node = fusedNode.dynamicCast<ONNXNodeWrapper>()->node;
+        opencv_onnx::AttributeProto* attr = node->add_attribute();
+        attr->set_name("scales");
+        attr->add_ints(scaleH);
+        attr->add_ints(scaleW);
+    }
+
+private:
+    int scaleH, scaleW;
+};
+
+class UpsampleSubgraph : public Subgraph
+{
+public:
+    UpsampleSubgraph() : scaleH(1), scaleW(1)
+    {
+        int input = addNodeToMatch("");
+
+        int scaleHW = addNodeToMatch("Scales", input);
+        int scaleNC = addNodeToMatch("Constant");
+        int cast1 = addNodeToMatch("Cast", scaleHW);
+
+        int shape = addNodeToMatch("Shape", input);
+        int slice = addNodeToMatch("Slice", shape);
+        int cast2 = addNodeToMatch("Cast", slice);
+
+        int inpDiv = addNodeToMatch("Div", cast1, cast2);
+        int inpConcat2 = addNodeToMatch("Concat", scaleNC, inpDiv);
+
+        addNodeToMatch("Upsample", input, inpConcat2);
+        setFusedNode("Upsample", input);
+    }
+
+    virtual bool match(const Ptr<ImportGraphWrapper>& net, int nodeId,
+                       std::vector<int>& matchedNodesIds,
+                       std::vector<int>& targetNodesIds) CV_OVERRIDE
+    {
+        if (Subgraph::match(net, nodeId, matchedNodesIds, targetNodesIds))
+        {
+            Ptr<ImportNodeWrapper> scales = net->getNode(matchedNodesIds[0]);
+            opencv_onnx::NodeProto* node = scales.dynamicCast<ONNXNodeWrapper>()->node;
+            opencv_onnx::AttributeProto attr = node->attribute(1);
+            CV_Assert(attr.name() == "scales");
+            CV_Assert(attr.ints_size() == 2);
+            scaleH = attr.ints(0);
+            scaleW = attr.ints(1);
+            return true;
+        }
+        return false;
+    }
+
+    virtual void finalize(const Ptr<ImportGraphWrapper>&,
+                          const Ptr<ImportNodeWrapper>& fusedNode,
+                          std::vector<Ptr<ImportNodeWrapper> >& inputs) CV_OVERRIDE
+    {
+        opencv_onnx::NodeProto* node = fusedNode.dynamicCast<ONNXNodeWrapper>()->node;
+        opencv_onnx::AttributeProto* attr_height = node->add_attribute();
+        attr_height->set_name("height_scale");
+        attr_height->set_i(scaleH);
+        opencv_onnx::AttributeProto* attr_width = node->add_attribute();
+        attr_width->set_name("width_scale");
+        attr_width->set_i(scaleW);
+    }
```

Comment:
Looks like finalize part is the same for all the cases. What do you think about creating a base class which implements `finalize` in this way:
```cpp
    virtual void finalize(const Ptr<ImportGraphWrapper>&,
                          const Ptr<ImportNodeWrapper>& fusedNode,
                          std::vector<Ptr<ImportNodeWrapper> >& inputs) CV_OVERRIDE
    {
        opencv_onnx::NodeProto* node = fusedNode.dynamicCast<ONNXNodeWrapper>()->node;
        opencv_onnx::AttributeProto* attr = node->add_attribute();
        attr->set_name("scales");
        attr->add_ints(scaleH);
        attr->add_ints(scaleW);
    }
```
so all the subgraphs will just override `match` and have different constructors?

Reply:
Yeah, I was thinking the same. just wanted to have your view. I will add that.